### ISSUE TITLE
prometheus-config-reloader has insufficient CPU resources

### ISSUE THREAD

**Issue by alejandrox1** [coreos/prometheus-operator#2575]
**What happened?**
I keep on getting constant cpu throttling alerts about the prometheus-config-reloader. i.e.,
```
37% throttling of CPU in namespace monitoring for container 
prometheus-config-reloader in pod prometheus-prometheus-operator-prometheus-0
```

**Did you expect to see some different?**
CPU limits for the prometheus-config-reloader should be configurable.
Looking at the official Helm chart there is a parameter `prometheusOperator.configReloaderCpu` that if set will add the following flag to the prometheus-operator pod:
```
 - --config-reloader-cpu={{ .Values.prometheusOperator.configReloaderCpu }}
```
https://github.com/helm/charts/blob/master/stable/prometheus-operator/templates/prometheus-operator/deployment.yaml#L51

However, this flag only affects the CPU limit for the configmap-reloader container (`quay.io/coreos/configmap-reload:v0.0.1`) not for the prometheus-config-reloader container (`quay.io/coreos/prometheus-config-reloader:v0.29.0`):
```
- args:
    - --log-format=logfmt
    - --reload-url=http://127.0.0.1:9090/-/reload
    - --config-file=/etc/prometheus/config/prometheus.yaml.gz
    - --config-envsubst-file=/etc/prometheus/config_out/prometheus.env.yaml
    command:
    - /bin/prometheus-config-reloader
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    image: quay.io/coreos/prometheus-config-reloader:v0.29.0
    imagePullPolicy: IfNotPresent
    name: prometheus-config-reloader
    resources:
      limits:
        cpu: 50m
        memory: 50Mi
      requests:
        cpu: 50m
        memory: 50Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/prometheus/config
      name: config
    - mountPath: /etc/prometheus/config_out
      name: config-out
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: prometheus-operator-prometheus-token-sqf2h
      readOnly: true
  - args:
    - --webhook-url=http://127.0.0.1:9090/-/reload
    - --volume-dir=/etc/prometheus/rules/prometheus-prometheus-operator-prometheus-rulefiles-0
    image: quay.io/coreos/configmap-reload:v0.0.1
    imagePullPolicy: IfNotPresent
    name: rules-configmap-reloader
    resources:
      limits:
        cpu: 100m
        memory: 25Mi
      requests:
        cpu: 100m
        memory: 25Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/prometheus/rules/prometheus-prometheus-operator-prometheus-rulefiles-0
      name: prometheus-prometheus-operator-prometheus-rulefiles-0
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: prometheus-operator-prometheus-token-sqf2h
      readOnly: true
  dnsPolicy: ClusterFirst
  hostname: prometheus-prometheus-operator-prometheus-0
  nodeName: ip-10-0-1-212.us-west-2.compute.internal
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    fsGroup: 2000
    runAsNonRoot: true
    runAsUser: 1000
  serviceAccount: prometheus-operator-prometheus
  serviceAccountName: prometheus-operator-prometheus
  subdomain: prometheus-operated
  terminationGracePeriodSeconds: 600
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
```

**How to reproduce it (as minimally and precisely as possible)**:
To see how the changes only affect the configmap-reloader container you can add this to the values.yaml file for the Helm chart
```
prometheusOperator:
  configReloaderCpu: 100m
```

**Environment**
Ran it on EKS, k8s v1.11.8

* Prometheus Operator version:
Helm chart version: 5.2.0
Prometheus operator version: 0.29.0

In summary, there should be a way of specifying resource limits for the prometheus-config-reloader container. It seems like the `config-reloader-cpu` flag for the prometheus operator does not properly set the limits for the `prometheus-config-reloader` and it only works with the `rules-configmap-reloader` container.
Thank you very much for looking at this.

**Comment by brancz** [comment]
What you're looking for has already been implemented in https://github.com/coreos/prometheus-operator/pull/2466. It just hasn't been released yet.

**Comment by JessieAMorris** [comment]
So, when's this looking to be released? The last release was back in Feb, we're now 3 months since then.

**Comment by metalmatze** [comment]
We're currently quite busy with the just announced OpenShift 4 release, but once we have some time, we should get another release out. Sorry and thanks for being patient.

**Comment by dimm0** [comment]
0.30.0 released 3 days ago, seems like with the fix, but still no docs on how to disable limits..

**Comment by alejandrox1** [comment]
Thank you for all the work you do @metalmatze @dimm0 
I'll test the latest release on a cluster and will come back to close this

**Comment by dimm0** [comment]
I'm not working on it :) thanks to @metalmatze 
And it seems like we still need to do something to disable the limits... I can be wrong. Testing it now as well

**Comment by JessieAMorris** [comment]
I authored the changes to allow people to change limits for all of the config reloader ones. To the best of my knowledge there's not a flag to disable limits for those containers entirely.

**Comment by dimm0** [comment]
https://github.com/coreos/prometheus-operator/releases the 0.30.0 release has:
`[FEATURE] Allow disabling limits on sidecars (#2560)`

**Comment by JessieAMorris** [comment]
Good catch. It looks like if you set the value to `0` it will disable the limits.

**Comment by dimm0** [comment]
It sets those to 0 in result yaml.. And leaves the sidecar ones

How hard is it to remove those from the repo?.. :( It's causing multiple problems

**Comment by brancz** [comment]
It's a flag to pass to the prometheus operator, not in the yaml.

**Comment by alejandrox1** [comment]
updating my prev comment: thank you @JessieAMorris for working on this.
update on the issue: been testing the v0.30 release and #2560 works as intended. So I'll close my issue.
Thank you everyone for keeping up on this 😄
